### PR TITLE
Support setting values on final fields

### PR DESCRIPTION
**Pull request by JakubGorski**

### CODE REVIEW

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. I will comment on the different commits

File: jOOR/src/main/java/org/joor/Reflect.java [hunk 1]
```diff
@@ -211,6 +211,34 @@ public Reflect set(String name, Object value) throws ReflectException {
     }
 
     /**
+     * Set a final field value.
+     * <p>
+     * Version of {@link Field#set(Object, Object)} only for changing final variables. If the
+     * wrapped object is a {@link Class}, then this will set a value to a static
+     * member field. If the wrapped object is any other {@link Object}, then
+     * this will set a value to an instance member field.
+     * For restrictions of usage check: http://bit.ly/changing_final_fields_java
```
Comment (lukaseder) [inline jOOR/src/main/java/org/joor/Reflect.java]: For security reasons, we shouldn't have URL shortened URLs in the Javadoc. This should be replaced by the actual URL:
http://stackoverflow.com/questions/3301635/change-private-static-final-field-using-java-reflection

File: jOOR/src/main/java/org/joor/Reflect.java [hunk 2]
```diff
@@ -211,6 +211,34 @@ public Reflect set(String name, Object value) throws ReflectException {
...
+     * member field. If the wrapped object is any other {@link Object}, then
+     * this will set a value to an instance member field.
+     * For restrictions of usage check: http://bit.ly/changing_final_fields_java
+     *
+     * @param name The final field name
+     * @param value The new field value
+     * @return The same wrapped object, to be used for further reflection.
+     * @throws ReflectException if any reflection exception occurred.
+     */
+    public Reflect setFinal(String name, Object value) throws ReflectException {
```
Comment (lukaseder) [inline jOOR/src/main/java/org/joor/Reflect.java]: I'd prefer if the existing `set()` method be modified to adopt the new behaviour.

**Comment by JakubGorski** [comment]
The only reason I had for implementing the new method was to protect users from their own mistakes - not let them change the final variables unless they consciously use the new method. But I agree such approach may not be needed.

**Comment by lukaseder** [comment]
Thanks for the update. I see your point, although, jOOR is a library that allows users to do precisely these things, including setting `private` members (things that probably won't work with Java 9 and/or a proper `SecurityManager`).

Since these things are possible to do with the reflection API, they should be possible to do with jOOR, too.

**Comment by lukaseder** [comment]
Nice idea with the nested properties in `testFinalFieldAdvanced()`, btw

**Comment by lukaseder** [comment]
Will also add a fat disclaimer in the Javadoc about this:
http://pveentjer.blogspot.co.at/2017/01/final-static-boolean-jit.html

**Pull request merged**